gnucash/import-export/import-main-matcher.c

```diff
@@ -1352,26 +1354,116 @@ void gnc_gen_trans_list_add_trans_with_ref_id (GNCImportMainMatcher *gui, Transa
+        import_trans_account = xaccSplitGetAccount (gnc_import_TransInfo_get_fsplit (transaction_info));
+        all_accounts = g_list_prepend (all_accounts, import_trans_account);
+        import_trans_time = xaccTransGetDate (gnc_import_TransInfo_get_trans (transaction_info));
+        min_time = MIN (min_time, import_trans_time);
+        max_time = MAX (max_time, import_trans_time);
+
+        valid = gtk_tree_model_iter_next (model, &iter);
+    }
 
-        gnc_import_TransInfo_init_matches (transaction_info,
-                                           gui->user_settings);
+    // Make a query to find splits with the right accounts and dates.
+    qof_query_set_book (query, gnc_get_current_book());
+    xaccQueryAddAccountMatch (query, all_accounts,
+                              QOF_GUID_MATCH_ANY, QOF_QUERY_AND);
+    xaccQueryAddDateMatchTT (query,
+                             TRUE, min_time - match_date_hardlimit * 86400,
+                             TRUE, max_time + match_date_hardlimit * 86400,
+                             QOF_QUERY_AND);
+    list_element = qof_query_run (query);
+    g_list_free (all_accounts);
+    return list_element;
+}
+
+static void
+gnc_gen_trans_list_finalize (GNCImportMainMatcher *gui)
+{
+    GtkTreeModel* model = gtk_tree_view_get_model (gui->view);
+    GtkListStore* store = GTK_LIST_STORE (model);
+    GtkTreeIter iter;
+    GNCImportMatchInfo *selected_match;
+    gboolean match_selected_manually;
+    Account *importaccount;
+    GList* list_element = NULL;
+    GList* iter_2 = NULL;
+    Query *query;
+    gint match_date_hardlimit = gnc_import_Settings_get_match_date_hardlimit (gui->user_settings);
+    gint display_threshold = gnc_import_Settings_get_display_threshold (gui->user_settings);
+    double fuzzy_amount = gnc_import_Settings_get_fuzzy_amount (gui->user_settings);
+    gboolean valid;
+
+    query = qof_query_create_for (GNC_ID_SPLIT);
+    // Gather the list of splits that could match one of the imported transactions
+    list_element = create_list_of_potential_matches (model, query, match_date_hardlimit);
+
+    // For each imported transaction, go through list_element, select the ones that have the
+    // right account, and evaluate how good a match they are.
+    valid = gtk_tree_model_get_iter_first (model, &iter);
+    while(valid)
+    {
+        GNCImportTransInfo* transaction_info;
+        Account *importaccount;
+        gtk_tree_model_get (model, &iter,
+                            DOWNLOADED_COL_DATA, &transaction_info,
+                            -1);
+        importaccount = xaccSplitGetAccount (gnc_import_TransInfo_get_fsplit (transaction_info));
+
+        iter_2 = list_element;
+        while (iter_2 != NULL)
+        {
+            // One issue here is that the accounts may not match since the query was
+            // called for all accounts so we need to check here.
+            Account* split_account = xaccSplitGetAccount(iter_2->data);
```

Comment:
Do you have a hunch as to whether that's a significant enough gain to warrant the added complexity? Actually it's not that much more complex, I'll give it a shot.

Reply:
The speed up would depend on how many accounts and how many transactions. With a single account there'd obviously be no speed up at all, in fact a slight slowdown because of the extra indirection; similarly if there were only a few transactions the speedup wouldn't be noticeable. On the other hand 5 accounts with 100 transactions each would get a 5N speedup where N is the number of items in the listmodel.